contracts/liquidity-protection/LiquidityProtection.sol

```diff
@@ -639,42 +632,147 @@ contract LiquidityProtection is ILiquidityProtection, Utils, Owned, ReentrancyGu
             }
-            _networkV3.migrateLiquidity{ value: value }(
-                removedPos.reserveToken,
-                provider,
+
+            // lock network tokens for the caller
+            _networkToken.safeTransfer(address(_wallet), delta);
+            _lockTokens(provider, delta);
+        }
+
+        // if the contract still holds network tokens, burn them
+        uint256 networkBalance = _networkToken.balanceOf(address(this));
+        if (networkBalance > 0) {
+            _burnNetworkTokens(removedPos.poolToken, networkBalance);
+        }
+    }
+
+    /**
+     * @dev migrates a set of position lists to v3
+     *
+     * Requirements:
+     *
+     * - the caller must be the owner of all of the positions
+     */
+    function migratePositions(MigrationUnit[] calldata migrationUnits) external nonReentrant {
+        uint256 length = migrationUnits.length;
+        for (uint256 i = 0; i < length; ++i) {
+            _migratePositions(migrationUnits[i]);
+        }
+    }
+
+    /**
+     * @dev migrates a list of positions to v3
+     *
+     * Requirements:
+     *
+     * - the caller must be the owner of all of the positions
+     */
+    function _migratePositions(MigrationUnit calldata migrationUnit) internal {
+        IDSToken poolToken = migrationUnit.poolToken;
+        IReserveToken reserveToken = migrationUnit.reserveToken;
+
+        (Fraction memory removeSpotRate, Fraction memory removeAverageRate) = _reserveTokenRates(
+            poolToken,
+            reserveToken
+        );
+
+        // verify rate deviation as early as possible in order to reduce gas-cost for failing transactions
+        _verifyRateDeviation(removeSpotRate.n, removeSpotRate.d, removeAverageRate.n, removeAverageRate.d);
+
+        uint256 poolAmount = 0;
+        uint256 reserveAmount = 0;
+        uint256 targetAmount = 0;
+
+        uint256 length = migrationUnit.positionIds.length;
+        for (uint256 i = 0; i < length; ++i) {
+            Position memory removedPos = _removePosition(msg.sender, migrationUnit.positionIds[i], PPM_RESOLUTION);
+            require(
+                removedPos.poolToken == poolToken && removedPos.reserveToken == reserveToken,
+                "ERR_INVALID_MIGRATION_UNIT"
+            );
+
+            // get the pool token amount
+            poolAmount = poolAmount.add(removedPos.poolAmount);
+
+            // get the reserve token amount
+            reserveAmount = reserveAmount.add(removedPos.reserveAmount);
+
+            // get the various rates between the reserves upon adding liquidity and now
+            PackedRates memory packedRates = _packRates(
+                removedPos.reserveRateN,
+                removedPos.reserveRateD,
+                removeSpotRate,
+                removeAverageRate
+            );
+
+            // get the target token amount
+            targetAmount = targetAmount.add(
```

Comment:
is there any value in trying to find out the aggregated IL and then calculating the target amount only once, as if it was all a big single position? any significant gas savings?

Reply:
That's not so possible, since different positions may have different impermanent losses and/or different protection levels attached to them.
However, a part of function `_removeLiquidityTargetAmount` does seem to be identical for all positions - a call to function `_poolTokenRate`, which yields 2 external calls.
So we can split function `_removeLiquidityTargetAmount` in a similar manner to how we've split function `_packRates` (called right before it), and perform these 2 external calls once for all positions.

Reply:
protection levels are accelerated to 100% for all of them but I think you're right that it might just be too complicated to normalize them based on the IL/weight of each :) fair enough

Reply:
The only significant part remaining in `_removeLiquidityTargetAmount` is:
```
uint256 minProtectionDelay = _settings.minProtectionDelay();
uint256 maxProtectionDelay = _settings.maxProtectionDelay();
```
We can replace it with constant values (which will impact both `migratePositions` and `removePosition`).
@yudilevi @lbeder : WDYT?

Reply:
NM, refactored the code to get rid of these two external calls for `migratePositions` only.